File: typescript/src/verifier/adapters/agentreceipts.ts
/**
 * agent-receipts adapter - W3C-VC AgentReceipt, Ed25519 over strict RFC 8785 JCS.
 * A port of the Python oracle's `verifier/oracle/adapters/agentreceipts.py`.
 *
 * An AgentReceipt is a W3C Verifiable Credential 2.0 envelope. The signature is
 * Ed25519 over the strict JCS bytes of the receipt with the `proof` member
 * removed, signed DIRECTLY. `proof.type` is `Ed25519Signature2020` and
 * `proof.proofValue` is multibase `u` (base64url, no padding).
 *
 * Uses `jcsRfc8785` (UTF-16 key sort, ECMAScript numbers), NOT the AERF/ACTA
 * `jcs` dialect. The signing key resolves from `proof.verificationMethod` through
 * the shared DID resolver.
 *
 * Chain rule: `credentialSubject.chain.previous_receipt_hash` is `"sha256:"` +
 * hex(SHA-256(JCS(predecessor without proof))). Genesis carries the field present
 * and explicitly `null`.
 */

import {
  FormatAdapter,
  type AxisCheck,
  type ChainStep,
  type KeyProvider,
  type SignatureMaterial,
} from "../adapter.js";
import { jcsRfc8785 } from "../canonical.js";
import { sha256Hex } from "../crypto.js";
import { resolveEd25519Key } from "../did.js";

/** SHA-256 hash prefix the chain field carries before the hex digest. */
const SHA256_PREFIX = "sha256:";

function signable(doc: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(doc)) {
    if (k !== "proof") out[k] = v;
  }
  return out;
}

/** Decode a multibase `u` (base64url, no padding) proofValue to raw bytes. */
function multibaseUDecode(value: string): Uint8Array {
  if (!value || value[0] !== "u") {
    throw new Error("proofValue is not multibase 'u' (base64url) encoded");
  }
  const body = value.slice(1).replace(/-/g, "+").replace(/_/g, "/");
  const padded = body + "=".repeat((-body.length % 4 + 4) % 4);
  return new Uint8Array(Buffer.from(padded, "base64"));
}

function asRecord(v: unknown): Record<string, unknown> {
  return v !== null && typeof v === "object" && !Array.isArray(v) ? (v as Record<string, unknown>) : {};
}

function chainOf(doc: Record<string, unknown>): Record<string, unknown> {
  const sub = doc.credentialSubject;
  if (sub !== null && typeof sub === "object" && !Array.isArray(sub)) {
    const chain = (sub as Record<string, unknown>).chain;
    return asRecord(chain);
  }
  return {};
}

export class AgentReceiptsAdapter extends FormatAdapter {
  readonly name = "agentreceipts";

  detect(doc: Record<string, unknown>): boolean {
    const types = doc.type;
    const proof = doc.proof;
    if (!Array.isArray(types) || !types.includes("AgentReceipt")) return false;
    return (
      proof !== null &&
      typeof proof === "object" &&
      !Array.isArray(proof) &&
      typeof (proof as Record<string, unknown>).proofValue === "string"
    );
  }

  extractSignature(doc: Record<string, unknown>): SignatureMaterial {
    const proof = asRecord(doc.proof);
    let sig: Uint8Array;
    try {
      sig = multibaseUDecode((proof.proofValue as string) ?? "");
    } catch {
      sig = new Uint8Array(0);
    }
    return { sig, alg: "Ed25519", kid: (proof.verificationMethod as string) ?? "" };
  }

  resolveKey(
    doc: Record<string, unknown>,
    keyProvider: KeyProvider,
  ): readonly [Uint8Array | null, string] {
    const kid = (asRecord(doc.proof).verificationMethod as string) ?? "";
    return resolveEd25519Key(kid, keyProvider);
  }

  signingInput(doc: Record<string, unknown>): Uint8Array {
    // Strict-JCS bytes of the receipt minus proof, signed directly (no pre-hash).
    return jcsRfc8785(signable(doc));
  }

  chainStep(doc: Record<string, unknown>): ChainStep {
    const chain = chainOf(doc);
    const present = "previous_receipt_hash" in chain;
    const prev = chain.previous_receipt_hash;
    const isGenesis = present && (prev === null || prev === undefined);
    return {
      prevField: (prev as string | null) ?? null,
      isGenesis,
      recompute: (pred) => SHA256_PREFIX + sha256Hex(jcsRfc8785(signable(pred))),
    };
  }

  schema(doc: Record<string, unknown>): AxisCheck {
    const required = [
      "@context",
      "id",
      "type",
      "version",
      "issuer",
      "issuanceDate",
      "credentialSubject",
      "proof",
    ];
    const missing = required.filter((f) => !(f in doc));
    if (missing.length > 0) {
      return ["FAIL", `missing required VC fields: ${missing.join(",")}`];
    }
    const types = doc.type;
    if (!Array.isArray(types) || !types.includes("VerifiableCredential") || !types.includes("AgentReceipt")) {
      return ["FAIL", "type must include VerifiableCredential and AgentReceipt"];
    }
    if (asRecord(doc.proof).type !== "Ed25519Signature2020") {
      return ["FAIL", "proof.type must be Ed25519Signature2020"];
    }
    const issuer = doc.issuer;
    const issuerDid =
      issuer !== null && typeof issuer === "object" && !Array.isArray(issuer)
        ? (issuer as Record<string, unknown>).id
        : issuer;
    const vmDid = ((asRecord(doc.proof).verificationMethod as string) ?? "").split("#")[0];
    if (!vmDid || vmDid !== issuerDid) {
      return ["FAIL", "proof.verificationMethod is not controlled by issuer (signing-key DID != issuer DID)"];
    }
    const chain = chainOf(doc);
    if (!("previous_receipt_hash" in chain)) {
      return ["FAIL", "chain.previous_receipt_hash must be present (null on genesis), never omitted"];
    }
    return ["PASS", "required VC fields present; type AgentReceipt; chain link well-formed"];
  }
}

File: python/src/asqav/verifier/oracle/adapters/agentreceipts.py
"""agent-receipts adapter - W3C-VC AgentReceipt, Ed25519 over strict RFC 8785 JCS.

An AgentReceipt is a W3C Verifiable Credential 2.0 envelope: top-level ``@context``,
``id``, ``type`` (``["VerifiableCredential", "AgentReceipt"]``), ``issuer``,
``issuanceDate``, ``credentialSubject``, and a ``proof``. The signature is Ed25519
over the strict JCS bytes of the receipt with the ``proof`` member removed, signed
DIRECTLY (no pre-hash) per the agent-receipts signing rule. The ``proof.type`` is
``Ed25519Signature2020`` and ``proof.proofValue`` is multibase ``u`` (base64url, no
padding).

This adapter uses ``jcs_rfc8785`` (UTF-16 key sort, ECMAScript numbers), NOT the
AERF/ACTA ``jcs`` dialect: the agent-receipts producers sign true strict-JCS bytes and
the upstream canonicalization vectors pin that. The signing key is resolved from
``proof.verificationMethod`` through the shared DID resolver - did:key inline, every
other method from an injected map (the oracle never fetches a DID document).

Chain rule: ``credentialSubject.chain.previous_receipt_hash`` is ``"sha256:"`` +
hex(SHA-256(JCS(predecessor without proof))). Genesis carries the field present and
explicitly ``null`` (the field MUST always be present; ``null`` is not the same as
omitting it) - the opposite of the AERF omit-on-genesis rule.
"""
from __future__ import annotations

import base64
from typing import Any

from ..adapter import ChainStep, FormatAdapter, SignatureMaterial
from ..canonical import jcs_rfc8785
from ..core import sha256_hex
from ..did import resolve_ed25519_key

#: SHA-256 hash prefix the chain field carries before the hex digest.
_SHA256_PREFIX = "sha256:"


def _signable(doc: dict) -> dict:
    """The receipt with ``proof`` removed - the exact object the signature covers."""
    return {k: v for k, v in doc.items() if k != "proof"}


def _multibase_u_decode(value: str) -> bytes:
    """Decode a multibase ``u`` (base64url, no padding) proofValue to raw bytes."""
    if not value or value[0] != "u":
        raise ValueError("proofValue is not multibase 'u' (base64url) encoded")
    body = value[1:]
    return base64.urlsafe_b64decode(body + "=" * (-len(body) % 4))


def _chain(doc: dict) -> dict:
    sub = doc.get("credentialSubject")
    return sub.get("chain", {}) if isinstance(sub, dict) else {}


class AgentReceiptsAdapter(FormatAdapter):
    """agent-receipts AgentReceipt - Ed25519 over strict RFC 8785 JCS, proof excluded."""

    name = "agentreceipts"

    def detect(self, doc: dict) -> bool:
        types = doc.get("type")
        proof = doc.get("proof")
        if not isinstance(types, list) or "AgentReceipt" not in types:
            return False
        return isinstance(proof, dict) and isinstance(proof.get("proofValue"), str)

    def extract_signature(self, doc: dict) -> SignatureMaterial:
        proof = doc.get("proof", {})
        try:
            sig = _multibase_u_decode(proof.get("proofValue", ""))
        except (ValueError, base64.binascii.Error):
            sig = b""
        return SignatureMaterial(sig=sig, alg="Ed25519", kid=proof.get("verificationMethod", ""))

    def resolve_key(self, doc: dict, key_provider: Any) -> tuple[bytes | None, str]:
        kid = doc.get("proof", {}).get("verificationMethod", "")
        return resolve_ed25519_key(kid, key_provider)

    def signing_input(self, doc: dict) -> bytes:
        # Strict-JCS bytes of the receipt minus proof, signed directly (no pre-hash).
        return jcs_rfc8785(_signable(doc))

    def chain_step(self, doc: dict) -> ChainStep:
        chain = _chain(doc)
        present = "previous_receipt_hash" in chain
        prev = chain.get("previous_receipt_hash")
        is_genesis = present and prev is None
        return ChainStep(
            prev_field=prev,
            is_genesis=is_genesis,
            recompute=lambda pred: _SHA256_PREFIX + sha256_hex(jcs_rfc8785(_signable(pred))),
        )

    def schema(self, doc: dict) -> tuple[str, str]:
        """Structural check; returns ``(result, note)``.

        Issuer binding: the strict ``verificationMethod`` DID == ``issuer.id`` check is
        intentionally stricter than the spec, which leaves controller indirection to DID
        resolution the oracle does not perform. It is fail-closed and revisited when
        DID-document resolution lands.
        """
        required = (
            "@context",
            "id",
            "type",
            "version",
            "issuer",
            "issuanceDate",
            "credentialSubject",
            "proof",
        )
        missing = [f for f in required if f not in doc]
        if missing:
            return "FAIL", f"missing required VC fields: {','.join(missing)}"
        types = doc.get("type")
        if not isinstance(types, list) or "VerifiableCredential" not in types or "AgentReceipt" not in types:
            return "FAIL", "type must include VerifiableCredential and AgentReceipt"
        if doc.get("proof", {}).get("type") != "Ed25519Signature2020":
            return "FAIL", "proof.type must be Ed25519Signature2020"
        issuer = doc.get("issuer")
        issuer_did = issuer.get("id") if isinstance(issuer, dict) else issuer
        vm_did = doc.get("proof", {}).get("verificationMethod", "").split("#")[0]
        if not vm_did or vm_did != issuer_did:
            # did:key carries its key in-receipt; bind to issuer or anyone self-signs as a victim
            return "FAIL", "proof.verificationMethod is not controlled by issuer (signing-key DID != issuer DID)"
        chain = _chain(doc)
        if "previous_receipt_hash" not in chain:
            return "FAIL", "chain.previous_receipt_hash must be present (null on genesis), never omitted"
        return "PASS", "required VC fields present; type AgentReceipt; chain link well-formed"
